File: aidl_server/src/main/java/com/mobile/aidl_server/service/AidlService.java
package com.mobile.aidl_server.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import com.mobile.aidl_client.Book;
import com.mobile.aidl_client.BookManager;

import java.util.ArrayList;
import java.util.List;

public class AidlService extends Service {
	
	private static final String TAG = "AidlService";
	
	@Override
	public void onCreate() {
		super.onCreate();
		Book book = new Book();
		book.setName("Android AIDL 测试通过");
		book.setContent("Android AIDL内容");
		mBooks.add(book);
	}
	
	@Override
	public IBinder onBind(Intent intent) {
		Log.i(TAG, String.format("on bind,intent = %s", intent.toString()));
		return mBookManager;
	}
	
	//包含Book对象的list
	private List<Book> mBooks = new ArrayList<>();
	
	//由AIDL文件生成的BookManager
	private final BookManager.Stub mBookManager = new BookManager.Stub() {
		@Override
		public List<Book> getBooks() throws RemoteException {
			synchronized (this) {
				Log.i(TAG, "invoking getBooks() method , now the list is : " + mBooks.toString());
				if (mBooks != null) {
					return mBooks;
				}
				return new ArrayList<>();
			}
		}
		
		@Override
		public void addBook(Book book) throws RemoteException {
			synchronized (this) {
				if (mBooks == null) {
					mBooks = new ArrayList<>();
				}
				if (book == null) {
					Log.i(TAG, "Book is null in In");
					book = new Book();
				}
				//尝试修改book的参数，主要是为了观察其到客户端的反馈
				book.setContent("Android AIDL内容");
				if (!mBooks.contains(book)) {
					mBooks.add(book);
				}
				//打印mBooks列表，观察客户端传过来的值
				Log.i(TAG, "invoking addBooks() method , now the list is : " + mBooks.toString());
			}
		}
	};
	
}

File: app_aidl_server/src/main/java/com/phone/aidl/app_aidl_server/AidlService.kt
package com.phone.aidl.app_aidl_server

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.os.RemoteException
import com.phone.aidl.aidl.Book
import com.phone.aidl.aidl.BookController
import com.phone.library_base.manager.LogManager
import java.util.ArrayList

class AidlService : Service() {

    private var mBookList: MutableList<Book> = ArrayList()

    override fun onCreate() {
        super.onCreate()
        initData()
    }

    private fun initData() {
        val book1 = Book("封神榜")
        val book2 = Book("公主小妹")
        val book3 = Book("仙剑奇侠传")
        val book4 = Book("飘")
        val book5 = Book("茶花女")
        val book6 = Book("解忧杂货铺")
        val book7 = Book("活着")
        val book8 = Book("三生三世十里桃花")
        mBookList.add(book1)
        mBookList.add(book2)
        mBookList.add(book3)
        mBookList.add(book4)
        mBookList.add(book5)
        mBookList.add(book6)
        mBookList.add(book7)
        mBookList.add(book8)
    }

    private val mStub: BookController.Stub = object : BookController.Stub() {
        @Throws(RemoteException::class)
        override fun getInt(): Int {
            return mBookList.size
        }

        @Throws(RemoteException::class)
        override fun getString(): String {
            return mBookList[0].name ?: ""
        }

        @Throws(RemoteException::class)
        override fun getBookList(): List<Book> {
            return mBookList
        }

        @Throws(RemoteException::class)
        override fun addBook(book: Book?) {
            if (book != null) {
                mBookList.add(book)
            } else {
                LogManager.i("ruxing", "接收到了一个空对象 Inout")
            }
        }
    }

    override fun onBind(intent: Intent?): IBinder {
        return mStub
    }


}